src/core/crypto/crypto_platform.cpp

```diff
@@ -0,0 +1,505 @@
+                                            size_t         aContextSize,
+                                            const uint8_t *aInfo,
+                                            uint16_t       aInfoLength,
+                                            uint8_t *      aOutputKey,
+                                            uint16_t       aOutputKeyLength)
+{
+    otError error = OT_ERROR_NONE;
+
+#if !OPENTHREAD_RADIO
+    HmacSha256        hmac;
+    HmacSha256::Hash  hash;
+    uint8_t           iter = 0;
+    uint16_t          copyLength;
+    HmacSha256::Hash *prk = static_cast<HmacSha256::Hash *>(aContext);
+
+    VerifyOrExit(aContextSize >= sizeof(HmacSha256::Hash), error = OT_ERROR_FAILED);
+
+    // The aOutputKey is calculated as follows [RFC5889]:
+    //
+    //   N = ceil( aOutputKeyLength / HashSize)
+    //   T = T(1) | T(2) | T(3) | ... | T(N)
+    //   aOutputKey is first aOutputKeyLength of T
+    //
+    // Where:
+    //   T(0) = empty string (zero length)
+    //   T(1) = HMAC-Hash(PRK, T(0) | info | 0x01)
+    //   T(2) = HMAC-Hash(PRK, T(1) | info | 0x02)
+    //   T(3) = HMAC-Hash(PRK, T(2) | info | 0x03)
+    //   ...
+
+    while (aOutputKeyLength > 0)
+    {
+        Key cryptoKey;
+
+        cryptoKey.Set(prk->GetBytes(), sizeof(HmacSha256::Hash));
+        hmac.Start(cryptoKey);
+
+        if (iter != 0)
+        {
+            hmac.Update(hash);
+        }
+
+        hmac.Update(aInfo, aInfoLength);
+
+        iter++;
+        hmac.Update(iter);
+        hmac.Finish(hash);
+
+        copyLength = (aOutputKeyLength > sizeof(hash)) ? sizeof(hash) : aOutputKeyLength;
+
+        memcpy(aOutputKey, hash.GetBytes(), copyLength);
+        aOutputKey += copyLength;
+        aOutputKeyLength -= copyLength;
+    }
+
+    error = OT_ERROR_NONE;
+#else
+    OT_UNUSED_VARIABLE(aContext);
+    OT_UNUSED_VARIABLE(aContextSize);
+    OT_UNUSED_VARIABLE(aInfo);
+    OT_UNUSED_VARIABLE(aInfoLength);
+    OT_UNUSED_VARIABLE(aOutputKey);
+    OT_UNUSED_VARIABLE(aOutputKeyLength);
+
+    ExitNow(error = OT_ERROR_NOT_IMPLEMENTED);
+#endif
+
+exit:
+    return error;
+}
+
+OT_TOOL_WEAK otError otPlatCryptoHkdfExtract(void *             aContext,
+                                             size_t             aContextSize,
+                                             const uint8_t *    aSalt,
+                                             uint16_t           aSaltLength,
+                                             const otCryptoKey *aKey)
+{
+    otError error = OT_ERROR_FAILED;
+
+#if !OPENTHREAD_RADIO
+    HmacSha256        hmac;
+    Key               cryptoKey;
+    HmacSha256::Hash *prk = static_cast<HmacSha256::Hash *>(aContext);
+
+    VerifyOrExit(aContextSize >= sizeof(HmacSha256::Hash), error = OT_ERROR_FAILED);
+
+    cryptoKey.Set(aSalt, aSaltLength);
+
+    // PRK is calculated as HMAC-Hash(aSalt, aInputKey)
+    hmac.Start(cryptoKey);
+
+    hmac.Update(aKey->mKey, aKey->mKeyLength);
```

Comment:
@hemanth-silabs I think we may have an issue (potenrtially incorrect behavior) in the `otPlatCryptoHkdfExtract()` function.

We are assuming here that `aKey` which is an `otCryptoKey` is always given as a literal key. 

However, in `KeyManager` when we use this we convert from `NetworkKey` and may end up using a `KeyRef` and not the literal key.

```c++
void KeyManager::ComputeTrelKey(uint32_t aKeySequence, Mac::Key &aKey)
{
    Crypto::HkdfSha256 hkdf;
    uint8_t            salt[sizeof(uint32_t) + sizeof(kHkdfExtractSaltString)];
    Crypto::Key        cryptoKey;

#if OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE
    cryptoKey.SetAsKeyRef(mNetworkKeyRef);
#else
    cryptoKey.Set(mNetworkKey.m8, NetworkKey::kSize);
#endif

    Encoding::BigEndian::WriteUint32(aKeySequence, salt);
    memcpy(salt + sizeof(uint32_t), kHkdfExtractSaltString, sizeof(kHkdfExtractSaltString));

    hkdf.Extract(salt, sizeof(salt), cryptoKey);
    hkdf.Expand(kTrelInfoString, sizeof(kTrelInfoString), aKey.m8, Mac::Key::kSize);
}
```

I think we should change the code here to get the literal key from the `otCryptoKey` if it happens to be a `KeyRef`. 

May be good to even add this as a helper method `ExtractKey()` in `Crypto::Key` class so we can use in other situations as well.

Reply:
I can help fix this. I will again make a copy of current branch on my fork and add a commit on top for this. Will post it later here.

Reply:
@hemanth-silabs here is the commits for the above change.
- Branch/commits -> https://github.com/abtink/openthread/commits/abtink/psa-2

There are two commits:

[**Fix `otPlatCryptoHkdfExtract()` to export the key if it is `KeyRef`**](https://github.com/abtink/openthread/commit/2f51543caa3bb18049bc437cd40c463ff8f1916d)

Second commit some smaller enhancments:

[**Misc. enhancement**](https://github.com/abtink/openthread/commit/6b1a4eac32e68e6ec415c34c87f09353e8385f03)

- Use `kError{Error}` constant and `Error` in `crypto_platform.cpp`
- Remove unused `CryptoType` enum in `mac_types.hpp`
- Remove unused `otCryptoType` enum in `platfomr/crypto.hpp`
  (we still have the `otRadioKeyType` in `radio.h`)
- Remove `otCryptoKeyUsage` enum name and mention that
  `OT_CRYPTO_KEY_USAGE_*` values are (bit)-flags.
- Update comment for `OT_CRYPTO_KEY_USAGE_ENCRYPT` value.

Can you please take a look to see all changes are ok, and help cherry-pick them on to the PR branch?  Thanks.

Reply:
@abtink that is not a bug. Basically you will either support mbedtls or PSA. If you support mbedtls you won't be using key refs and in PSA you won't be using literal keys. The default implementation `otPlatCryptoHkdfExtract()` is for mbedTLS and we should not be receiving key refs here, in fact you can't with mbedtls. For PSA we are passing key ref and the vendor implementation of crypto abstraction will not need to get the literal key as it works on key refs.

Reply:
Also for mbedtls we will have to implement HKDF using HMAC, as there are no direct APIs for HKDF. But in PSA we have APIs to perform HKDF operations and the software despot need to do the HMAC calculations, so no literal strings needed to feed the HMAC operations.

Reply:
@abtink are you OK with just cherrypicking the second commit?

Reply:
> that is not a bug. Basically you will either support mbedtls or PSA. If you support mbedtls you won't be using key refs and in PSA you won't be using literal keys. The default implementation of `otPlatCryptoHkdfExtract()` is for mbedTLS and we should not be receiving key refs here, in fact you can't with mbedtls. 

I think it may be possible for a platform to want to use `mbedTLS` along with some platform-specific secure key storage and therefore the `KeyRef` mechanism for passing keys. In such a case, the platform layer may provide its own `otPlatCryptoImportKey()`, `otPlatCryptoExportKey()`, etc, but not necessarily provide `otPlatCryptoHkdfExtract()` and use the weak default implementation we provide in OT.  I think we need to handle this case. 
- Does this make sense? Or do we want to assume `OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE` always means that platform should also use PSA and provide all the `otPlatCrypto` APIs?

Here are the option I can suggest:
- If we want to keep `otPlatCryptoHkdfExtract()` as is, then I think we need to add it under `#if !OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE` to make sure it cannot be used with `KeyRef`.
- Or we update it to handle both literal key or `KeyRef` (which is the change in the commit I added). 

I think second option may be better and more flexible. What do you think?

Reply:
@abtink unless there are any plans to implement secure-key storage in OT core, I think we won't need to support Key Refs in crypto_platform.cpp. I say this because, if any platform wants to continue with mbedTLS but have a platform specific implementation of secure key storage, it is possible for them to do this by implementing a crypto platform abstraction. In these cases, crypto_platform.cpp won't be used at all.

So I am in favour of option 1. I suggest modifying the current implementation of crypto_platform.cpp, and changing `#if !OPENTHREAD_RADIO` to #if !OPENTHREAD_RADIO && ! OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE`

But if you feel we need to support option 2, we will need to extend such support to other cases too (like AES). What do you think?